### ISSUE TITLE
get_actor_status_info query hangs for ever

### ISSUE THREAD

**Issue by iazehaf** [ansys/pyoptislang#181]
### 🔍 Before submitting the issue

- [X] I have searched among the existing issues
- [X] I am using a Python virtual environment

### 🐞 Description of the bug

I tried to run the Hook-Optimization use case. When I try to monitor the progress using PyOptiSLang the following command never returns (it hangs forever), therefore preventing the program to continue:
`actor_status_info = osl.get_osl_server().get_actor_status_info(node.uid, hid)`

This query takes place in a while loop which monitors the project state:

```
while project_state not in ["FINISHED", "ABORTED"]:
        project_state = osl.project.get_status()
        for node_props in osl.project._get_project_tree():
            if not node_props["is_root"]:
                node = osl.project.root_system.find_node_by_uid(node_props["uid"], search_depth=-1)
                states_ids = node.get_states_ids()
                for hid in states_ids:
                    actor_status_info = osl.get_osl_server().get_actor_status_info(node.uid, hid)
```

At some point of the analysis, the last statement of the code above will cause the program to loop forever. 

### 📝 Steps to reproduce

Run the Hook-Optimization use case using the following code:

```
osl = Optislang(
    project_path="hook_optimization.opf",
    loglevel="INFO",
    reset=True,
    shutdown_on_finished=True,
    import_project_properties_file="hook_optimization.json",
    ini_timeout=300
)

osl.start(wait_for_started=True, wait_for_finished=False)

project_state = osl.project.get_status()

while project_state not in ["FINISHED", "ABORTED"]:
  project_state = osl.project.get_status()
  for node_props in osl.project._get_project_tree():
      if not node_props["is_root"]:
          node = osl.project.root_system.find_node_by_uid(node_props["uid"], search_depth=-1)
          states_ids = node.get_states_ids()
          for hid in states_ids:
              actor_status_info = osl.get_osl_server().get_actor_status_info(node.uid, hid)
```

### 💻 Which operating system are you using?

Windows

### 📀 Which ANSYS version are you using?

23R1

### 🐍 Which Python version are you using?

3.8

### 📦 Installed packages

```shell
Package                                       Version      Editable project location
--------------------------------------------- ------------ -----------------------------------------------
aiofiles                                      23.2.1
aiomysql                                      0.2.0
aiosqlite                                     0.19.0
annotated-types                               0.5.0
ansi2html                                     1.8.0
ansys-api-platform-instancemanagement         1.0.0
ansys-optislang-core                          0.4.0
ansys-platform-instancemanagement             1.1.2
anyio                                         4.0.0
asgiref                                       3.7.2
async-timeout                                 4.0.3
backoff                                       2.2.1
bottle                                        0.12.25
bson                                          0.5.10
cachelib                                      0.9.0    
certifi                                       2023.7.22
cffi                                          1.15.1   
charset-normalizer                            3.2.0    
click                                         8.1.7    
clr-loader                                    0.2.6    
colorama                                      0.4.6    
coloredlogs                                   15.0.1   
cryptography                                  41.0.3   
dash                                          2.13.0   
dash-bootstrap-components                     1.5.0    
dash-core-components                          2.0.0    
dash-daq                                      0.5.0    
dash-extensions                               0.1.13   
dash-html-components                          2.0.0    
dash-table                                    5.0.0    
dash-uploader                                 0.6.0    
debugpy                                       1.8.0
Deprecated                                    1.2.14
EditorConfig                                  0.12.3
exceptiongroup                                1.1.3
fastapi                                       0.100.1
fire                                          0.5.0
Flask                                         2.2.5
Flask-Caching                                 2.0.2
googleapis-common-protos                      1.60.0
greenlet                                      2.0.2
grpcio                                        1.58.0
grpcio-health-checking                        1.48.2
h11                                           0.14.0
httpcore                                      0.16.3
httpx                                         0.23.3
humanfriendly                                 10.0
idna                                          3.4
importlib-metadata                            6.0.1
itsdangerous                                  2.1.2
Jinja2                                        3.1.2
jsbeautifier                                  1.14.9
MarkupSafe                                    2.1.3
more-itertools                                9.1.0
nest-asyncio                                  1.5.7
networkx                                      3.1
numpy                                         1.24.4
opentelemetry-api                             1.18.0
opentelemetry-exporter-otlp                   1.18.0
opentelemetry-exporter-otlp-proto-common      1.18.0
opentelemetry-exporter-otlp-proto-grpc        1.18.0
opentelemetry-exporter-otlp-proto-http        1.18.0
opentelemetry-instrumentation                 0.39b0
opentelemetry-instrumentation-asgi            0.39b0
opentelemetry-instrumentation-fastapi         0.39b0
opentelemetry-instrumentation-flask           0.39b0
opentelemetry-instrumentation-httpx           0.39b0
opentelemetry-instrumentation-logging         0.39b0
opentelemetry-instrumentation-wsgi            0.39b0
opentelemetry-proto                           1.18.0
opentelemetry-sdk                             1.18.0
opentelemetry-semantic-conventions            0.39b0
opentelemetry-util-http                       0.39b0
optislang-dash-lib                            0.2.1
packaging                                     23.1
pandas                                        2.0.3
pip                                           23.2.1
plotly                                        5.16.1
protobuf                                      3.20.3
proxy-tools                                   0.1.0
psutil                                        5.9.5
pycparser                                     2.21
pydantic                                      2.3.0
pydantic_core                                 2.6.3
pydantic-settings                             2.0.3
PyMySQL                                       1.1.0
pyreadline3                                   3.4.1
pyshortcuts                                   1.9.0
python-dateutil                               2.8.2
python-dotenv                                 1.0.0
python-json-logger                            2.0.7
python-multipart                              0.0.6
pythonnet                                     3.0.2
pytz                                          2023.3.post1
pywebview                                     4.3.3
pywin32                                       306
PyYAML                                        6.0.1
requests                                      2.31.0
retrying                                      1.3.4
rfc3986                                       1.5.0
setuptools                                    67.8.0
six                                           1.16.0
sniffio                                       1.3.0
SQLAlchemy                                    2.0.20
starlette                                     0.27.0
tenacity                                      8.2.3
termcolor                                     2.3.0
typing_extensions                             4.7.1
tzdata                                        2023.3
urllib3                                       2.0.4
uvicorn                                       0.23.2
Werkzeug                                      2.2.3
wheel                                         0.40.0
wrapt                                         1.15.0
zipp                                          3.16.2
```

**Comment by rfahlberg** [comment]
@iazehaf Just to be sure: Are you using the above snippet in that way in SAF oSL project monitoring? Including looping over all hid in states_ids and fetching get_actor_status_info for each hid? This should not be necessary at least for the  diplayed results per item in the project tree, as this should only display the status info for a specific selected hid.

**Comment by rfahlberg** [comment]
> Hi, could you please provide `hook_optimization.opf` and `hook_optimization.json` so I can try to reproduce? @iazehaf
> 
> Just to clarify, it hangs on the execution of `actor_status_info = osl.get_osl_server().get_actor_status_info(node.uid, hid)` line, not the for loop (so it cannot be case with hundreds of hids, where the last for loop just takes a lot of time)?
> 
> Is the server still running while execution of `actor_status_info = osl.get_osl_server().get_actor_status_info(node.uid, hid)` hangs?

@PanekOndrej This does not seem to be an isolated PyOptiSLang issue, but with the underlying TCP socket communication. Sometimes, requests or responses just doesn't seem to be delivered/received. I'm not sure, if there is anything that can be done on either optiSLang or PyOptiSLang side. I think, it can always happen, that something get's lost. IMO, what we should do, is to be aware of that and improve the timeout handling and error reporting, like e.g.:
- provide multiple configurable timeouts for different purposes (e.g. one for basic requests which are expected to finish quickly, one for more complex requests, ... ; separate query and command timeouts; ...)
- possibility to finetune timeouts per request / API call (e.g. optional timeout argument, ...)

**Comment by PanekOndrej** [comment]
I see, both options:
 - optional argument `timeout` in all commands and queries 
 - 2 different general timeouts for queries & commands 
 are definitely possible. 

We just have to be careful not to overcomplicate it. We already have `timeout` and `ini_timeout`...

What about resending queries automatically, if first attempt fails? This could reduce unnecessary errors caused by lost data.

**Comment by ekrja** [comment]
> @PanekOndrej This does not seem to be an isolated PyOptiSLang issue, but with the underlying TCP socket communication. Sometimes, requests or responses just doesn't seem to be delivered/received. I'm not sure, if there is anything that can be done on either optiSLang or PyOptiSLang side. I think, it can always happen, that something get's lost. IMO, what we should do, is to be aware of that and improve the timeout handling and error reporting, like e.g.:
> 
> * provide multiple configurable timeouts for different purposes (e.g. one for basic requests which are expected to finish quickly, one for more complex requests, ... ; separate query and command timeouts; ...)
> * possibility to finetune timeouts per request / API call (e.g. optional timeout argument, ...)

@rfahlberg I think that data loss should not occur with TCP socket communication, because TCP is reliable protocol with retransmission and error detection.

**Comment by rfahlberg** [comment]
> What about resending queries automatically, if first attempt fails? This could reduce unnecessary errors caused by lost data.

@PanekOndrej @ekrja @widerschein
I'm currently investigating those timeouts a bit. At least in my tests, I'm experiencing those only seldom (like 1 timeout per 10k queries) and only on Windows. I was not able to reproduce them on Linux at all. The actual Error I always get is _TimeoutError ('timed out')_, which seems to by a pure OS level socket timeout (Python doc: states "Raised when a system function timed out at the system level. Corresponds to errno [ETIMEDOUT](https://docs.python.org/3/library/errno.html#errno.ETIMEDOUT).")

I could also not find any hint, that there is anything wrong in optiSLang server socket communication. When i get those timeouts, there is no trace of anything fishy in the optiSLang instance. To blocked/stray query processing threads, no errors, etc..

I cannot put my finger on it, but my gut feeling is, that really something on OS socket communication level gets amiss.

tl/dr I'd say, implicitly resending timed out requests is a vaiable option. But maybe really only for requests with this exact Exception (_TimeoutError ('timed out')_)?

**Comment by rfahlberg** [comment]
> @iazehaf Just to be sure: Are you using the above snippet in that way in SAF oSL project monitoring? Including looping over all hid in states_ids and fetching get_actor_status_info for each hid? This should not be necessary at least for the diplayed results per item in the project tree, as this should only display the status info for a specific selected hid.

@iazehaf Can you please comment on this?

**Comment by iazehaf** [comment]
@rfahlberg, Sorry for the delay. Yes the code snippet posted in the issue description is exactly what has been implemented in SAF oSL project monitoring.

**Comment by rfahlberg** [comment]
> @rfahlberg, Sorry for the delay. Yes the code snippet posted in the issue description is exactly what has been implemented in SAF oSL project monitoring.

@iazehaf OK, then this needs to be adapted in SAF. Do not loop over all hids for each node. Display info for only one single hid, selected by user.
